File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0901_Online_Stock_Span.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Stack;

public class Question_0901_Online_Stock_Span {
    private Stack<int[]> stack;

    public Question_0901_Online_Stock_Span() {
        this.stack = new Stack<>();
    }

    public int next(int price) {
        int res = 1;
        while (!stack.empty() && stack.peek()[0] <= price) {
            res += stack.peek()[1];
            stack.pop();
        }
        stack.push(new int[]{price, res});
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0901_Online_Stock_Span.cc
#include <stack>

using namespace std;

class StockSpanner {
 public:
  StockSpanner() {
  }

  int next(int price) {
    int res = 1;
    while (!s.empty() && s.top().first <= price) {
      res += s.top().second;
      s.pop();
    }
    s.push({price, res});
    return res;
  }

 private:
  stack<pair<int, int>> s;
};
